src/coordinator/ts/m3db/block/ns_series.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"math"
+
+	"github.com/m3db/m3db/src/coordinator/block"
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/ts"
+	"github.com/m3db/m3x/ident"
+)
+
+// ConsolidatedNSBlock is a single block for a given timeseries and namespace
+// which contains all of the necessary SeriesIterators so that consolidation can
+// happen across namespaces
+type ConsolidatedNSBlock struct {
+	ID              ident.ID
+	Namespace       ident.ID
+	Bounds          block.Bounds
+	SeriesIterators encoding.SeriesIterators
+}
+
+type consolidatedNSBlockIter struct {
+	m3dbIters        []encoding.SeriesIterator
+	bounds           block.Bounds
+	seriesIndex, idx int
+	lastDP           ts.Datapoint
+}
+
+type consolidatedNSBlockIters []*consolidatedNSBlockIter
+
+// Next moves to the next item
+func (c *consolidatedNSBlockIter) Next() bool {
+	c.idx++
+	// NB(braskin): this is inclusive of the last step in the iterator
+	indexTime, err := c.bounds.TimeForIndex(c.idx)
+	if err != nil {
```

Comment:
Feels a little strange to return no next when we have an error. Can you add a comment here as to the motivation for this?

Reply:
edit; looked at TimeForIndex and looks like it might need a little tweaking to resolve my problems here; would you be able to change it to something like:

```golang
// TimeForIndex returns the start time for a given index assuming a uniform step size
func (b Bounds) TimeForIndexWithinBounds(idx int) (time.Time, bool) {
	step := b.StepSize
	t := b.Start.Add(time.Duration(idx) * step)
	if t.After(b.End) {
		return time.Time{}, false
	}

	return t, false
}
```

Then we can avoid relying on `errors` for execution

Reply:
I'm gonna leave this for now. I'll add a comment though.